### ISSUE TITLE
accessing gradients

### ISSUE THREAD

**Issue by ostwind** [tensorflow/federated#756]
Hi TFF team!

I'm working on a Tensorflow Federated tutorial/viz that aims to draw an analogy between SGD and user gradients within Federated Learning. To do this, I want to access the gradients on each client device, and emit them (they may be outputted to file or returned by a federated_computation). Is this possible and what would be the best way to do this? 

I tried:  modifying the FC tutorial part 2 to return the gradient along with the new weights, 
and modifying the tf gradient update procedure to output the gradient to file. Both are stopped by the tff syntactic sugar, so to speak.

Thank you and please continue on this wonderful project.

**Comment by jkr26** [comment]
Hi @ostwind,

Thanks for your interest in TFF, and the compliment! 

This is definitely *doable* in TFF, although yes, it is probably a little unnatural. Generally, TFF wants to make it difficult to inspect values on a single client, e.g. by modeling the federated type placed at CLIENTS as a single multiset, instead of via addressable individual clients.

Given this, I can imagine at least three distinct possibilities here, depending on exactly what you are looking for:

First: If you are just looking for the sum of the computed gradients per client, you can simply call [`tff.federated_collect`](https://www.tensorflow.org/federated/api_docs/python/tff/federated_collect), which collects models from each client and places them on the server. This can be instrumented right in the FC tutorial, like:
```python
SERVER_FLOAT_TYPE = tff.FederatedType(tf.float32, tff.SERVER, all_equal=True)

@tff.federated_computation(
    SERVER_MODEL_TYPE, SERVER_FLOAT_TYPE, CLIENT_DATA_TYPE)
def grab_all_models(model, learning_rate, data):
  return tff.federated_collect(
      tff.federated_map(
          local_train,
          [tff.federated_broadcast(model),
           tff.federated_broadcast(learning_rate),
           data]))
```
And calling:
```python
models = grab_all_models(initial_model, 0.1, federated_train_data)
```
will result in a list of containers on the server representing the models at the end of training on each client. You can compute the difference between these models and the original model, resulting in the sum of the gradients seen on a per-client basis. One thing to note here is that the data container TFF will spit out on the server is a custom container type; I would recommend accessing the numpy array elements underneath. This is a bit of a rough edge in TFF right now, resulting in necessity of calls like:
```
[x.bias - initial_model['bias'] for x in models]
```
Generally, this edge is handled by [some conversion function](https://github.com/tensorflow/federated/blob/master/tensorflow_federated/python/research/utils/checkpoint_utils_test.py#L42), but for this example this is unnecessary.

By the way, if you do end up running the line above, you will see that each client learned its digit-wise bias, as expected in the FC tutorial. Which IMO is pretty cool!

Second: If you are looking for literally every gradient each client has produced during training, you will again need to `tff.federated_collect` these gradients, but first you will need to pack them into either a [`tf.data.Dataset`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset), which TFF will recognize as a [`tff.SequenceType`](https://www.tensorflow.org/federated/api_docs/python/tff/SequenceType). I think this would exercise a codepath that is relatively uncovered in TFF, though there are [some tests](https://github.com/tensorflow/federated/blob/master/tensorflow_federated/python/core/api/computations_test.py#L356) which should cover returning Datasets, so we should support this case.

If this is what you're looking for, this may get a little more involved, so I won't go further for now.

Related: we are in the process of open-sourcing an unrolled, reference federated averaging implementation, which will probably be a better model to follow for either use case than the FC tutorial. I'll follow up when it's out and link to it.

Thanks again!
Keith

**Comment by jkr26** [comment]
Adding Zach as he is shepherding the reference FedAvg into the light

**Comment by ostwind** [comment]
Thanks @jkr26 for the comprehensive explanation. I opted for route 1 and it was straightforward to analyze the gradients now.  

> By the way, if you do end up running the line above, you will see that each client learned its digit-wise bias, as expected in the FC tutorial. Which IMO is pretty cool!

I was able to observe this as well. After plotting out the weights and bias gradient sums, each client's gradient converges toward the global model's gradient as training goes on (global model is the black trajectory, with 6 Fashion-MNIST classes shown, tutorial's model).      

Please let me know if the team has particular topics of interest. I hope to say something more about client data's resemblance to IID sampling with validation loss. (e.g. how different is FL from SGD in those conditions?)      

![pca3](https://user-images.githubusercontent.com/4082500/64317369-7e6a3380-cf85-11e9-86cc-77c5c4089c7b.png)

Thanks again! I will wait for federated averaging link but feel free to close this as it is resolved.

**Comment by jkr26** [comment]
Following up here, this is the implementation I was talking about:

https://github.com/tensorflow/federated/tree/master/tensorflow_federated/python/research/simple_fedavg

I think generally in FL, relationship to IID/non-IID is a quite interesting and important area of research, as is exploring the relationship you describe.

Closing issue. Thanks for working in TFF!

**Comment by rufengma** [comment]
> Thanks @jkr26 for the comprehensive explanation. I opted for route 1 and it was straightforward to analyze the gradients.
> 
> > By the way, if you do end up running the line above, you will see that each client learned its digit-wise bias, as expected in the FC tutorial. Which IMO is pretty cool!
> 
> I was able to observe this. After plotting out the weights and bias gradient sums, each client's gradient converges toward the global model's gradient as training goes on (global model is the black trajectory, with 6 Fashion-MNIST classes shown, tutorial's model).
> 
> Please let me know if the team has particular topics of interest. I hope to say something more about validation loss curve with respect to client data's resemblance to IID sampling. (e.g. how different is FL from SGD in those conditions?)
> 
> ![pca3](https://user-images.githubusercontent.com/4082500/64317369-7e6a3380-cf85-11e9-86cc-77c5c4089c7b.png)
> 
> Thanks again! I will wait for federated averaging link but feel free to close this as it is resolved.

Hi, I just have a. Do you call the federated_train function too? Or just called the grab_all_models? If I do not call the tff.federated_mean function, the weights are not updated.
